apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabTheorycraft/BuildConstaintCard.tsx

```diff
@@ -0,0 +1,143 @@
+import { StatIcon } from '@genshin-optimizer/gi-svgicons'
+import {
+  CardThemed,
+  CustomNumberInput,
+  CustomNumberInputButtonGroupWrapper,
+  DropdownButton,
+} from '@genshin-optimizer/ui-common'
+import { unit } from '@genshin-optimizer/util'
+import ClearIcon from '@mui/icons-material/Clear'
+import {
+  Box,
+  Button,
+  ButtonGroup,
+  CardHeader,
+  Divider,
+  ListItemIcon,
+  ListItemText,
+  MenuItem,
+  Stack,
+} from '@mui/material'
+import { useContext } from 'react'
+import { ArtifactStatWithUnit } from '../../../../Components/Artifact/ArtifactStatKeyDisplay'
+import type { MinTotalStatKey } from '../../../../Database/DataManagers/CharacterTCData'
+import { minTotalStatKeys } from '../../../../Database/DataManagers/CharacterTCData'
+import { CharTCContext } from './CharTCContext'
+import { useTranslation } from 'react-i18next'
+
+export function BuildConstaintCard({ disabled }: { disabled: boolean }) {
+  const { t } = useTranslation('page_character')
+  const {
+    charTC: {
+      optimization: { minTotal },
+    },
+    setCharTC,
+  } = useContext(CharTCContext)
+
+  return (
+    <CardThemed bgt="light">
+      <CardHeader title={t`tabTheorycraft.constraint.title`} />
+      <Divider />
+      <Box sx={{ p: 1 }}>
+        <Stack spacing={1}>
+          {Object.entries(minTotal).map(([k, v]) => (
+            <Selector key={k} statKey={k} value={v} disabled={disabled} />
+          ))}
+          <DropdownButton
+            disabled={disabled}
+            title={t`tabTheorycraft.constraint.add`}
+          >
+            {minTotalStatKeys.map((k) => (
+              <MenuItem
+                key={k}
+                disabled={Object.keys(minTotal).includes(k)}
+                onClick={() =>
+                  setCharTC((charTC) => {
+                    charTC.optimization.minTotal[k] = 0
+                  })
+                }
+              >
+                <ListItemIcon>
+                  <StatIcon statKey={k} />
+                </ListItemIcon>
+                <ListItemText>
+                  <ArtifactStatWithUnit statKey={k} />
+                </ListItemText>
+              </MenuItem>
+            ))}
+          </DropdownButton>
+        </Stack>
+      </Box>
+    </CardThemed>
+  )
+}
+
+function Selector({
+  statKey,
+  value,
+  disabled,
+}: {
+  statKey?: MinTotalStatKey
+  value?: number
+  disabled: boolean
+}) {
+  const { setCharTC } = useContext(CharTCContext)
+  const unitStr = unit(statKey) || ' '
+  return (
+    <ButtonGroup size="small">
+      <CustomNumberInputButtonGroupWrapper sx={{ flexBasis: 30, flexGrow: 1 }}>
+        <CustomNumberInput
+          float={unitStr === '%'}
```

Comment:
Don't we need useCallback calls for the various event handlers in this component to prevent recalculating so often?

Reply:
useCallback is more used to prevent propagating rerenders in child component, e.g. passing in a memoized callback so that child doesnt rerender when the callback is updated. This doesn't apply here, because its a child component, and that when rerender happens (value changes, the function is regenerated), it doesnt cause/prevent more rerenders.

Reply:
Actually, the onChange can be memoized, which can prevent one layer of rerendering, but since its so close to the input, i'd argue if its worth the overhead of memorizing.